R/git.R

```diff
@@ -132,70 +132,70 @@ use_git_config <- function(scope = c("user", "project"), ...) {
-#' option: `options(usethis.protocol = "ssh")` or
-#' `options(usethis.protocol = "https")`.
+#'   * The strategy for creating `credentials` when none are given. See
+#'     [git2r_credentials()] for details.
+#' Two helper functions are relevant:
+#'   * `git_protocol()` returns the user's preferred protocol, if known, and,
+#'     otherwise, asks the user (interactive session) or defaults to SSH
+#'     (non-interactive session).
+#'   * `use_git_protocol()` allows the user to set the git protocol, which is
+#'     stored in the `usethis.protocol` option.
+#' Any interactive choice re: `protocol` comes with a reminder of how to set the
+#' protocol at startup by setting an option in `.Rprofile`:
+#' ```
+#' options(usethis.protocol = "ssh")
+#' ## or
+#' options(usethis.protocol = "https")
+#' ```
 #'
-#' @section Default credentials:
-#'
-#' The current git protocol is consulted to create default credentials when
-#' they are needed but left unspecified.
-#'
-#' For `protocol = "ssh"`, usethis passes `NULL` credentials straight through to
-#' git2r. This will do the right thing if you have the exact configuration
-#' expected by git2r: your public and private keys are in the default locations,
-#' `~/.ssh/id_rsa.pub` and `~/.ssh/id_rsa`, respectively, and your `ssh-agent`
-#' is configured to manage any associated passphrase, if relevant. Read more
-#' about SSH setup in [Happy Git](http://happygitwithr.com/ssh-keys.html),
-#' especially the [troubleshooting
-#' section](http://happygitwithr.com/ssh-keys.html#ssh-troubleshooting).
-#'
-#' For `protocol = "https"`, usethis attempts to use a GitHub personal access
-#' token (PAT), with preference for an `auth_token` that is passed explicitly.
-#' If that is `NULL`, the environment variables `GITHUB_PAT` and `GITHUB_TOKEN`
-#' are consulted, in that order. If a PAT is found, we make an HTTPS credential
-#' with [git2r::cred_user_pass()]. The PAT is sent as the password and dummy
-#' text is sent as the username (only the PAT is truly required).
-#'
-#' @param protocol "ssh" or "https". Defaults to the option `usethis.protocol`
-#'   and, if unset, to an interactive choice or, in non-interactive sessions,
-#'   "ssh".
+#' @param protocol Optional. Should be "ssh" or "https", if specified. Defaults
+#'   to the option `usethis.protocol` and, if unset, to an interactive choice
+#'   or, in non-interactive sessions, "ssh". `NA` triggers the interactive menu.
 #'
 #' @return "ssh" or "https"
 #' @export
 #'
 #' @examples
 #' \dontrun{
-#' use_git_protocol()
+#' ## consult the option and maybe get an interactive menu
+#' git_protocol()
+#'
+#' ## explicitly set the protocol
 #' use_git_protocol("ssh")
 #' use_git_protocol("https")
 #' }
-use_git_protocol <- function(protocol = NULL) {
-  if (!is.null(protocol)) {
-    stopifnot(
-      length(protocol) == 1,
-      tolower(protocol) %in% c("ssh", "https", NA)
+git_protocol <- function() {
```

Comment:
When I tackle the `pr_*()` stuff, this will presumably gain some smarts re: consulting the "current repo" if such a thing is well-defined. I think that will come into focus then.

Reply:
In the existing use cases (`use_github()` and `create_from_github()`) there is no existing local <--> remote relationship to consult re: protocol. But that will not be true for `pr_*()` functions.

Reply:
I think we might also be able to add a bit more smarts with @jeroen's credentials package — like if an ssh key exists, we could assume that the protocol should be ssh.

Reply:
Yeah I'm trying to do this in a way that will make fitting that package in a fairly pleasant task. It's on my radar.